### PR TITLE
Allow TreeTraversal to run individual queries in non-batch mode

### PR DESCRIPTION
**Pull request by masterleinad**
Related to https://github.com/arborx/ArborX/pull/542, https://github.com/arborx/ArborX/pull/784, and https://github.com/arborx/ArborX/pull/903.
This pull request allows creating a `TreeTraversal` object without specifying predicates and running the queries in non-batch mode using a `search()` member function. Since we are in a `Detail` namespace, there is no need to be backward-compatible.

### CODE REVIEW

**Comment by masterleinad** [comment]
Trying this in the BVH benchmark gives somewhat mixed results, see https://code.ornl.gov/ecpcitest/alexa/-/jobs/1927138.
```
BM_radius_callback_search<ArborX::BVH<Serial>>/1000/1000/10/1/0/0/2/manual_time_median                    +0.0055         +0.0055          3452          3471          3459          3478
BM_radius_callback_search<ArborX::BVH<Serial>>/10000/10000/10/1/0/0/2/manual_time_median                  +0.0463         +0.0463         39708         41546         39710         41550
BM_radius_callback_search<ArborX::BVH<Serial>>/100000/100000/10/1/0/0/2/manual_time_median                +0.1011         +0.1011        414191        456052        414136        455992
BM_radius_callback_search<ArborX::BVH<Serial>>/1000/1000/10/1/0/1/3/manual_time_median                    +0.0498         +0.0495          1753          1840          1761          1848
BM_radius_callback_search<ArborX::BVH<Serial>>/10000/10000/10/1/0/1/3/manual_time_median                  +0.1800         +0.1800         12476         14721         12482         14729
BM_radius_callback_search<ArborX::BVH<Serial>>/100000/100000/10/1/0/1/3/manual_time_median                +0.1787         +0.1786         95215        112227         95212        112220
BM_radius_callback_search<ArborX::BVH<OpenMP>>/1000/1000/10/1/0/0/2/manual_time_median                    -0.8553         -0.8444           674            97           682           106
BM_radius_callback_search<ArborX::BVH<OpenMP>>/10000/10000/10/1/0/0/2/manual_time_median                  -0.3657         -0.3620          1559           989          1570          1002
BM_radius_callback_search<ArborX::BVH<OpenMP>>/100000/100000/10/1/0/0/2/manual_time_median                -0.0261         +0.0083         10659         10380         10328         10413
BM_radius_callback_search<ArborX::BVH<OpenMP>>/1000/1000/10/1/0/1/3/manual_time_median                    -0.9084         -0.8959           625            57           634            66
BM_radius_callback_search<ArborX::BVH<OpenMP>>/10000/10000/10/1/0/1/3/manual_time_median                  -0.6674         -0.5906          1157           385           968           396
BM_radius_callback_search<ArborX::BVH<OpenMP>>/100000/100000/10/1/0/1/3/manual_time_median                -0.5546         +0.1590          6124          2728          2382          2761
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/0/2/manual_time_median                    -0.6594         -0.5967           444           151           489           197
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/0/2/manual_time_median                  -0.6442         -0.6231          1331           474          1388           523
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/0/2/manual_time_median                +2.1922         +1.9718          6323         20183          7031         20893
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/1/3/manual_time_median                    -0.7617         -0.6858           405            96           450           141
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/1/3/manual_time_median                  -0.8177         -0.7874          1198           218          1254           267
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/1/3/manual_time_median                -0.4627         -0.3820          3379          1816          4087          2526
```

**Comment by aprokop** [comment]
> Trying this in the BVH benchmark gives somewhat mixed results

This is likely because we ran sorted queries in Gitlab regression suite, but the current patch does not sort them. So, it's not comparing identical paths.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 1]
```diff
@@ -70,6 +70,11 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
     }
   }
 
+  TreeTraversal(BVH const &bvh, Callback const &callback)
+      : _bvh{bvh}
+      , _callback{callback}
+  {}
+
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:77]: Generally, I wonder if this should be the only constructor, and whether the `parallel_for` should live outside of the `TreeTraversal` (possibly in a free function).

**Comment by masterleinad** [comment]
> This is likely because we ran sorted queries in Gitlab regression suite, but the current patch does not sort them. So, it's not comparing identical paths.

Yes, I didn't mean to say that we were comparing the same code path. I merely wanted to express that we don't overly bad results when not sorting.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 1 above)
Comment (masterleinad) [reply src/details/ArborX_DetailsTreeTraversal.hpp:77]: In an earlier version, I had a run` member functions but then dropped to minimize changes. If we want to allow users to call the search in this non-batched way, i.e., inside of kernels, we should also provide a better interface than through `ArborX::Details::TreeTraversal`. I could imagine deriving a public (non-detail) class from this one or a free function.

Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp:77]: Wonder if we should provide
```c++
class BVH {
template<typename Predicate, typename Callback>
KOKKOS_FUNCTION void query(Predicate const& predicate, Callback const& callback);
}
```
or
```c++
template<typename BVH, typename Predicate, typename Callback>
KOKKOS_FUNCTION void query(BVH const& bvh, Predicate const& predicate, Callback const& callback);
```

File: benchmarks/bvh_driver/benchmark_registration.hpp [hunk 2]
```diff
@@ -196,6 +196,19 @@ struct CountCallback
...
+  Kokkos::View<int *, DeviceType> count_;
+
+  template <typename Query, typename Primitive>
+  KOKKOS_FUNCTION void operator()(Query const &query, Primitive const &) const
+  {
+    auto const i = ArborX::getData(query);
+    Kokkos::atomic_increment(&count_(i));
+  }
+};
+
```
Comment (aprokop) [inline benchmarks/bvh_driver/benchmark_registration.hpp:211]: Why not modify the existing `CountCallback`? Just switch the second argument from hardcoded `int` to `Value` template argument.

File: benchmarks/bvh_driver/benchmark_registration.hpp [hunk 3]
```diff
@@ -297,6 +310,52 @@ void BM_radius_callback_search(benchmark::State &state, Spec const &spec)
       spec.n_queries, benchmark::Counter::kIsIterationInvariantRate);
 }
 
+template <typename ExecutionSpace, class TreeType>
+void BM_radius_callback_search_kernel_query(benchmark::State &state,
```
Comment (aprokop) [inline benchmarks/bvh_driver/benchmark_registration.hpp:314]: I'd like to see the results before seeing if we need it. As all the threads are launched at the same time in the batched mode anyway, I don't see how this is different from the usual query, and what insight it gives us. If the results are not the same, it is a problem that needs to be fixed, as imho they should be.

File: src/ArborX_CrsGraphWrapper.hpp [hunk 4]
```diff
@@ -41,6 +42,16 @@ inline void query(Tree const &tree, ExecutionSpace const &space,
   Kokkos::Profiling::popRegion();
 }
 
+template <typename Tree, typename Predicate, typename Callback>
+KOKKOS_FUNCTION void kernel_query(Tree const &tree, Predicate const &predicate,
+                                  Callback const &callback)
+{
```
Comment (aprokop) [inline src/ArborX_CrsGraphWrapper.hpp:48]: Honestly, I'm not sure whether this is the right interface. The alternative would be something like
```c++
struct BoundingVolumeHierarchy {
  ...
  template<typename Predicate, typename Callback>
  auto query(Predicate const& predicate, Callback const& callback);
```
This has an advantage of maintaining close interface to what we have right now.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 5]
```diff
@@ -70,6 +70,11 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
     }
   }
 
+  KOKKOS_FUNCTION TreeTraversal(BVH const &bvh, Callback const &callback)
+      : _bvh{bvh}
+      , _callback{callback}
+  {}
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:76]: I don't know if the current form of `TreeTraversal` is ideal. Imho, we could move the launch loop into `BVH::query`. This way, we wouldn't need any constructors at all, and could make it a struct, or have a single constructor. 

Not confident, though.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 6]
```diff
@@ -88,7 +93,12 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
   KOKKOS_FUNCTION void operator()(int queryIndex) const
   {
     auto const &predicate = Access::get(_predicates, queryIndex);
+    search(predicate);
+  }
 
+  template <typename Query>
+  KOKKOS_FUNCTION void search(Query const &predicate) const
+  {
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:101]: Instead of introducing a function, we could only have a thread variant, and grab the query inside the parallel_loop.

File: benchmarks/bvh_driver/benchmark_registration.hpp (hunk 3 above)
Comment (masterleinad) [reply benchmarks/bvh_driver/benchmark_registration.hpp:314]: You can find results above. At this point, the only difference is in predicate sorting.

File: src/ArborX_CrsGraphWrapper.hpp (hunk 4 above)
Comment (masterleinad) [reply src/ArborX_CrsGraphWrapper.hpp:48]: This is what I had first but I thought we prefer non-class calls. It's easy to add so that the non-class version would call this one.

File: benchmarks/bvh_driver/benchmark_registration.hpp (hunk 3 above)
Comment (aprokop) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: > You can find results above. At this point, the only difference is in predicate sorting.

Sure, but that's what I'm asking. If we do the same thing, i.e. sort predicates, is there any difference?

File: src/ArborX_CrsGraphWrapper.hpp (hunk 4 above)
Comment (aprokop) [reply src/ArborX_CrsGraphWrapper.hpp:48]: > This is what I had first but I thought we prefer non-class calls.

Don't know if that's true. There's no preference in my head, and I don't think we ever had a discussion.

File: benchmarks/bvh_driver/benchmark_registration.hpp (hunk 3 above)
Comment (masterleinad) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: A100, without sorting
```
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/0/2/manual_time_median                        2538 us         7350 us           15 rate=3.94059M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/0/2/manual_time_median           2538 us         7351 us           15 rate=3.93997M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/0/2/manual_time_median                      2708 us         7516 us           15 rate=36.927M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/0/2/manual_time_median         2708 us         7516 us           15 rate=36.9247M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/0/2/manual_time_median                   13391 us        18197 us           15 rate=74.6772M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/0/2/manual_time_median      13373 us        18185 us           15 rate=74.7793M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/1/3/manual_time_median                        2485 us         7297 us           15 rate=4.02472M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/1/3/manual_time_median           2485 us         7297 us           15 rate=4.02467M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/1/3/manual_time_median                      2567 us         7375 us           15 rate=38.9593M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/1/3/manual_time_median         2567 us         7375 us           15 rate=38.9629M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/1/3/manual_time_median                    3628 us         8441 us           15 rate=275.632M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/1/3/manual_time_median       3627 us         8440 us           15 rate=275.685M/s
```
A100 with sorting
```
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/0/2/manual_time_median                         252 us          279 us           15 rate=39.6682M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/0/2/manual_time_median            146 us          173 us           15 rate=68.3766M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/0/2/manual_time_median                       537 us          570 us           15 rate=186.177M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/0/2/manual_time_median          317 us          345 us           15 rate=315.853M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/0/2/manual_time_median                    2584 us         2844 us           15 rate=387.052M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/0/2/manual_time_median       4881 us         5126 us           15 rate=204.894M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/1/3/manual_time_median                        5710 us        11246 us           15 rate=1.75143M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/1/3/manual_time_median           2860 us         8412 us           15 rate=3.49685M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/1/3/manual_time_median                      5819 us        11217 us           15 rate=17.1852M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/1/3/manual_time_median         2888 us         8330 us           15 rate=34.6241M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/1/3/manual_time_median                    1368 us         2029 us           15 rate=730.801M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/1/3/manual_time_median       1534 us         2202 us           15 rate=651.871M/s
```

Comment (aprokop) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: > A100 with sorting

When you say sorting, are they sorted only in the batched mode and not in the query mode?

I assume no. The first (unsorted) clearly indicates there's no difference, which is to be expected. If that's also true for both sorted, I see no point in modifying the benchmark at all.

Comment (masterleinad) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: Sorting should not have an impact on the batch-mode so the current performance testing setting allows us to compare sorted and unsorted results. Ultimately, I can duplicate any other test so that we have something before #902 and I chose the benchmark since the question about difference in performance came up.

Comment (aprokop) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: Comparing sorted and unsorted results is outside of the scope of this PR. We don't have a good handle on when sorting actually is good or not. See, for example, #474. The point of this PR is simply a refactoring and a new interface, in my opinion. Performance of the unsorted queries is completely orthogonal to that.

Comment (masterleinad) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: Right, but we don't monitor performance in the case of not sorting so we would kill two birds with one stone.
Let me duplicate another test in `tstQueryTreeCallbacks.cpp` anyway.

Comment (aprokop) [reply benchmarks/bvh_driver/benchmark_registration.hpp:301]: I think the only test that is required is the verification test that the results produced by a batched query and the individual calls in the loop are the same. It would have to use `TraversalPolicy` to guarantee that the sorting of the queries does not depend on the default.

File: src/ArborX_LinearBVH.hpp [hunk 7]
```diff
@@ -88,6 +88,13 @@ class BasicBoundingVolumeHierarchy
                   std::forward<View>(view), std::forward<Args>(args)...);
   }
 
+  template <typename Predicate, typename Callback>
+  KOKKOS_FUNCTION void kernel_query(Predicate const &predicate,
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:92]: Why not just `query`?

**Review by aprokop — COMMENTED** [review]
I think this PR should have 3 changes only:
- Introduce `BVH::query` for a single thread
  Get rid of `ArborX::query` in CrsGraph, and move its contents to this function
- The changes to `TreeTraversal` structures
  Add constructor, change `operator()` to take in a predicate instead of index, modify the parallel loops in the other constructors to grab a predicate
- A test verifying that using a single-thread interface produces the same result as the usual query

Ideally, we would also somehow indicate that `BVH::query` is experimental.

File: src/ArborX_LinearBVH.hpp (hunk 7 above)
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp:92]: Any prior known practice to prepend `kernel_`?

Comment (masterleinad) [reply src/ArborX_LinearBVH.hpp:92]: I'm open to other naming suggestions but that was the first one that came to mind.

**Comment by masterleinad** [comment]
>  change operator() to take in a predicate instead of index, modify the parallel loops in the other constructors to grab a predicate

We can't use a `KOKKOS_LAMBDA` in a constructor because `nvcc` doesn't like it, we need to use a call operator. To avoid selecting the wrong overload, I introduced a tag for the "normal" tree traversal. Not sure if this better than having a `search()` member function and leave the other member functions as-is, though.

**Comment by masterleinad** [comment]
@aprokop I tried to address your suggestions.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 8]
```diff
@@ -443,6 +461,12 @@ struct TreeTraversal<BVH, Predicates, Callback,
   KOKKOS_FUNCTION void operator()(int queryIndex) const
   {
     auto const &predicate = Access::get(_predicates, queryIndex);
+    search(predicate);
+  }
+
+  template <typename Query>
+  KOKKOS_FUNCTION void search(Query const &predicate) const
+  {
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:469]: This code is different from the spatial one. Is there a reason they should be?

File: test/tstQueryTreeCallbackKernelQuery.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <numeric>
+#include <vector>
+
+BOOST_AUTO_TEST_SUITE(KerneliQueryCallbacks)
+
+struct IntersectionCallback
+{
+  int query_index;
+  bool &success;
```
Comment (aprokop) [inline test/tstQueryTreeCallbackKernelQuery.cpp:25]: This is a gray area.

File: test/tstQueryTreeCallbackKernelQuery.cpp [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  template <typename Query, typename Value>
+  KOKKOS_FUNCTION void operator()(Query const &, Value const &value) const
+  {
+    ArborX::Box actual_box = value.bounding_volume;
+    ArborX::Point min = actual_box.minCorner();
+    if (query_index != min[0] || query_index != min[1] || query_index != min[2])
+      success = false;
+    else
+      success = true;
+  }
```
Comment (aprokop) [inline test/tstQueryTreeCallbackKernelQuery.cpp:36]: This seems fragile. Why not do
```suggestion
    success = (query_index == value.index);
```

File: test/tstQueryTreeCallbackKernelQuery.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include <ArborX.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+#include <numeric>
+#include <vector>
+
+BOOST_AUTO_TEST_SUITE(KerneliQueryCallbacks)
```
Comment (aprokop) [inline test/tstQueryTreeCallbackKernelQuery.cpp:20]:
```suggestion
BOOST_AUTO_TEST_SUITE(PerThread)
```

File: test/tstQueryTreeCallbackKernelQuery.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  using MemorySpace = typename DeviceType::memory_space;
+  using ExecutionSpace = typename DeviceType::execution_space;
+  using Tree = ArborX::BVH<MemorySpace>;
+
+  int const n = 10;
+  Kokkos::View<ArborX::Point *, DeviceType> points(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "points"), n);
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(0, n), KOKKOS_LAMBDA(int i) {
+        points(i) = {{(double)i, (double)i, (double)i}};
```
Comment (aprokop) [inline test/tstQueryTreeCallbackKernelQuery.cpp:51]: Why `double` and not `float`?

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 13]
```diff
@@ -63,15 +63,23 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
...
 
+  KOKKOS_FUNCTION TreeTraversal(BVH const &bvh, Callback const &callback)
+      : _bvh{bvh}
+      , _callback{callback}
+  {}
+
   struct OneLeafTree
   {};
 
+  struct OrdinaryTree
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeTraversal.hpp:80]: Perhaps `FullTree` instead of `OrdinaryTree`? Or `NontrivialTree`?

File: src/ArborX_LinearBVH.hpp [hunk 14]
```diff
@@ -91,6 +97,18 @@ class BasicBoundingVolumeHierarchy
                   std::forward<View>(view), std::forward<Args>(args)...);
   }
 
+  template <typename Predicate, typename Callback>
+  KOKKOS_FUNCTION void query(Experimental::PerThread,
+                             Predicate const &predicate,
+                             Callback const &callback) const
+  {
+    ArborX::Details::TreeTraversal<BasicBoundingVolumeHierarchy,
+                                   /* Dummy */ std::true_type, Callback,
```
Comment (aprokop) [inline src/ArborX_LinearBVH.hpp:106]:
```suggestion
                                   /* Predicates dummy */ std::true_type, Callback,
```

File: test/tstQueryTreeCallbackKernelQuery.cpp (hunk 10 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackKernelQuery.cpp]: That's fair. Changed.

File: test/tstQueryTreeCallbackKernelQuery.cpp (hunk 12 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackKernelQuery.cpp]: No good reason. Changed.

File: test/tstQueryTreeCallbackKernelQuery.cpp (hunk 9 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackKernelQuery.cpp]: Does this violate any of our semantics? Why do you call it "gray area"?

File: test/tstQueryTreeCallbackKernelQuery.cpp (hunk 11 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackKernelQuery.cpp]: Sure.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 8 above)
Comment (masterleinad) [reply src/details/ArborX_DetailsTreeTraversal.hpp]: I just removed the overloads for this variant for now.

File: src/ArborX_LinearBVH.hpp (hunk 14 above)
Comment (masterleinad) [reply src/ArborX_LinearBVH.hpp]: Sure.

File: test/tstQueryTreeCallbackKernelQuery.cpp (hunk 9 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackKernelQuery.cpp]: For me, being able to capture variables by reference is what makes this interface quite useful/powerful.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 8 above)
Comment (aprokop) [reply src/details/ArborX_DetailsTreeTraversal.hpp]: Why? Would be nice to have both spatial and nearest single-thread variants. Given that for example XGC asked us to find intersection, and if there's no intersection, find the closest.

Comment (masterleinad) [reply src/details/ArborX_DetailsTreeTraversal.hpp]: That's what I thought initially. I still added the implementation for `OrderedSpatial` and a corresponding test.
`Nearest` isn't quite clear to me, though, so I would omit it for now.

**Review by aprokop — APPROVED** [review]
LGTM

**Comment by aprokop** [comment]
One of the builds fail with a real error.

File: test/ArborX_EnableDeviceTypes.hpp.in [hunk 15]
```diff
@@ -14,8 +14,8 @@
 
 #include <Kokkos_Core.hpp>
 
-#include <tuple>
+#include <boost/mpl/list.hpp>
 
-using ARBORX_DEVICE_TYPES = std::tuple<@ARBORX_DEVICE_TYPES@>;
+using ARBORX_DEVICE_TYPES = boost::mpl::list<@ARBORX_DEVICE_TYPES@>;
```
Comment (aprokop) [inline test/ArborX_EnableDeviceTypes.hpp.in:20]: Please include `BoostTest_CUDA_clang_workarounds.hpp` instead.

Comment (masterleinad) [reply test/ArborX_EnableDeviceTypes.hpp.in]: Here you go!

**Comment by masterleinad** [comment]
HIP and SYCL were timing out.

File: src/details/ArborX_DetailsTreeTraversal.hpp (hunk 13 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTreeTraversal.hpp]: FWIW I don't like FullTree at all and think "ordinary" (or "regular") was better

File: src/ArborX_LinearBVH.hpp [hunk 16]
```diff
@@ -130,10 +148,11 @@ class BoundingVolumeHierarchy
   {}
 
   template <typename ExecutionSpace, typename Predicates, typename Callback>
-  void query(ExecutionSpace const &space, Predicates const &predicates,
-             Callback const &callback,
-             Experimental::TraversalPolicy const &policy =
-                 Experimental::TraversalPolicy()) const
+  std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value>
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:151]: Why did you add enable_if?

File: test/tstQueryTreeCallbackQueryPerThread.cpp [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+struct OrderedIntersectionCallback
+{
+  int query_index;
+  bool &success;
+
+  template <typename Query, typename Value>
+  KOKKOS_FUNCTION auto operator()(Query const &, Value const &value) const
+  {
+    success = (query_index == value.index);
+    return ArborX::CallbackTreeTraversalControl::early_exit;
```
Comment (dalg24) [inline test/tstQueryTreeCallbackQueryPerThread.cpp:76]: What is this testing?  Why are we doing early exit?

File: src/ArborX_LinearBVH.hpp (hunk 16 above)
Comment (masterleinad) [reply src/ArborX_LinearBVH.hpp:151]: Overload resolution picked the wrong overload.

File: test/tstQueryTreeCallbackQueryPerThread.cpp (hunk 17 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: The queries are executed in order of the distance. We are only interested in the first intersection that has zero distance.

File: src/ArborX_LinearBVH.hpp (hunk 16 above)
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp:151]: How is that so?  There is no ambiguity between the two.

File: test/tstQueryTreeCallbackQueryPerThread.cpp (hunk 17 above)
Comment (dalg24) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: Looks to me that the predicate geometry is a box that will only contain a single point.

File: src/ArborX_LinearBVH.hpp (hunk 16 above)
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp]: To elaborate why I care about this, arguments passed to the old BVH::query overload are checked via these `check_valid_*` function and the SFINAE would yield poor diagnostic and confuse users.

File: test/tstQueryTreeCallbackQueryPerThread.cpp (hunk 17 above)
Comment (dalg24) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: There is only one primitive that will match what am I missing

Comment (masterleinad) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: `OrderedIntersects` checks the distance for all primitives and executes them in order of increasing distance. For rays, this behavior corresponds to checking for intersections and that's why we chose the name. In this case, we would really process all primitives for all predicates (since all distances are finite).

File: test/tstQueryTreeCallbackQueryPerThread.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+  using MemorySpace = typename DeviceType::memory_space;
+  using ExecutionSpace = typename DeviceType::execution_space;
+  using Tree = ArborX::BVH<MemorySpace>;
+
+  int const n = 10;
+  Kokkos::View<ArborX::Point *, DeviceType> points(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "points"), n);
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(0, n), KOKKOS_LAMBDA(int i) {
+        points(i) = {{(float)i, (double)i, (double)i}};
```
Comment (dalg24) [inline test/tstQueryTreeCallbackQueryPerThread.cpp:92]: Why is this mixing double and float?!

Comment (masterleinad) [reply test/tstQueryTreeCallbackQueryPerThread.cpp]: I forgot to change all of them when Andrey wanted to change from `double` to `float`. It doesn't change anything, of course.

**Comment by masterleinad** [comment]
Retest this please.

File: test/tstQueryTreeCallbackQueryPerThread.cpp (hunk 17 above)
Comment (masterleinad) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: @dalg24 Did my explanation clarify why we need `early_exit` here?

Comment (dalg24) [reply test/tstQueryTreeCallbackQueryPerThread.cpp:76]: No it still does not make sense to me.  You are discussing some possible use of the callback.  I am looking at the one concrete use below and there is a single intersection so order does not matter.  Simplify your callback or change your test.
